python/cuspatial/cuspatial/core/geoseries.py

```diff
@@ -768,63 +775,174 @@ def contains_properly(self, other, align=True):
+        return CoversBinpred(self, other, align)()
+
+    def intersects(self, other, align=True):
+        """Compute the intersections of two GeoSeries.
+
+        Parameters
+        ----------
+        other
+            a cuspatial.GeoSeries
+        align=True
+            align the GeoSeries indexes before calling the binpred
+
+        Returns
+        -------
+        result : cudf.Series
+            A Series of boolean values indicating whether the geometries of
+            each row intersect.
+        """
+        if contains_only_points(self) and contains_only_points(other):
+            return cudf.Series(EqualsBinpred(self, other, align)())
+        else:
+            return IntersectsBinpred(self, other, align)()
+
+    def within(self, other, align=True):
+        """An object is said to be within other if at least one of its points
+        is located in the interior and no points are located in the exterior
+        of the other.
+
+        Parameters
+        ----------
+        other
+            a cuspatial.GeoSeries
+        align=True
+            align the GeoSeries indexes before calling the binpred
+
+        Returns
+        -------
+        result : cudf.Series
+            A Series of boolean values indicating whether each feature falls
+            within the corresponding polygon in the input.
+        """
+        if contains_only_points(self) and contains_only_points(other):
+            return cudf.Series(EqualsBinpred(self, other, align)())
+        else:
+            return WithinBinpred(self, other, align)()
+
+    def overlaps(self, other, align=True):
+        """Returns True for all aligned geometries that overlap other, else
+        False.
+
+        Geometries overlap if they have more than one but not all points in
+        common, have the same dimension, and the intersection of the
+        interiors of the geometries has the same dimension as the geometries
+        themselves.
+
+        Parameters
+        ----------
+        other
+            a cuspatial.GeoSeries
+        align=True
+            align the GeoSeries indexes before calling the binpred
+
+        Returns
+        -------
+        result : cudf.Series
+            A Series of boolean values indicating whether each geometry
+            overlaps the corresponding geometry in the input."""
+        # Overlaps has the same requirement as crosses.
+        if contains_only_points(self) and contains_only_points(other):
+            return cudf.Series([False] * len(self))
+        else:
+            return OverlapsBinpred(self, other, align=align)()
+
+    def crosses(self, other, align=True):
+        """Returns True for all aligned geometries that cross other, else
+        False.
+
+        Geometries cross if they have some but not all interior points in
+        common, have the same dimension, and the intersection of the
+        interiors of the geometries has the dimension of the geometries
+        themselves minus one.
```

Comment:
So if we have two polygons that overlap, do they also cross?  Confused by the "dimension of the geometries themselves minus 1". 

How about a line that crosses a polygon. The intersection has 1D geometry, right? But that is not one less than the dimension of both the geometries -- the line is 1D and the polygon is 2D.

Reply:
In A Small Set of Formal Topological Relationships Suitable for End User Interaction (SSFTR), $cross$ is defined as

$$$
< \lambda_1, cross \lambda_2> \iff dim(\lambda_1^o \cap \lambda_2^o) = (max(dim(\lambda_1^o), dim(\lambda_2^o))-2 \land (\lambda_1 \cap \lambda_2 \neq \lambda_1) \land (\lambda_1 \cap \lambda_2 \neq \lambda_2)
$$$

Reply:
Two polygons that overlap I think share an interior dimension, that is $$dim(\lambda_1^o \cap \lambda_2^o) = 2$$.

In order for two polygons to cross they have to share an overlap that is just a line, and it can't only be in the exterior. This is essentially impossible for two areas, so we may not see cases of polygon crossing.

`cross` looks like it applies primarily to linestrings, as the dimensionality of the interior overlap will always be 0 for linestring.cross(linestring) and 1 for `linestring.cross(polygon)` or `polygon.cross(linestring)`.

Reply:
Looking again at SSFTR I see that `cross` is applied when $max(dim(\lambda_1^o), dim(\lambda_2^o)) -1 = \dim(\lambda_1^o \cap \lambda_2^o)$ and `overlap` is applied when the dimension of the interior intersection equals the dimension of the interior.